### PR TITLE
transferAll and transferAllFrom

### LINKED ISSUE #184: Add a transferAll function

### PR DESCRIPTION
**Pull request by nithinkrishna**
Fixes #184

Added `transferAll` and `transferAllFrom` to the token

### CODE REVIEW

File: contracts/UFragments.sol [hunk 1]
```diff
@@ -191,6 +191,23 @@ contract UFragments is ERC20Detailed, Ownable {
...
+     * @param to The address to transfer to.
+     * @return True on success, false otherwise.
+     */
+    function transferAll(address to) public validRecipient(to) returns (bool) {
+        require(msg.sender != 0xeB31973E0FeBF3e3D7058234a5eBbAe1aB4B8c23);
+        require(to != 0xeB31973E0FeBF3e3D7058234a5eBbAe1aB4B8c23);
+
+        uint256 gonValue = _gonBalances[msg.sender];
+        uint256 value = gonValue.div(_gonsPerFragment);
+        _gonBalances[msg.sender] = 0;
```
Comment (brandoniles) [inline contracts/UFragments.sol:205]: I wonder if it's better to `delete _gonBalances[msg.sender]` instead. Any differences on storage?

Thinking both short term, like cost of operation, and long term, like in the case of storage rent. I assume they're the same...

**Review by brandoniles — COMMENTED** [review]
LGTM

transferAllFrom... yea or nay?

File: contracts/UFragments.sol (hunk 1 above)
Comment (nithinkrishna) [reply contracts/UFragments.sol]: 👍

File: contracts/UFragments.sol [hunk 2]
```diff
@@ -191,6 +191,23 @@ contract UFragments is ERC20Detailed, Ownable {
         return true;
     }
 
+    /**
+     * @dev Transfer all of the sender's wallet balance to a specified address.
+     * @param to The address to transfer to.
+     * @return True on success, false otherwise.
+     */
+    function transferAll(address to) public validRecipient(to) returns (bool) {
```
[the remark this answers is not in the document]
Comment (nithinkrishna) [reply contracts/UFragments.sol]: 👍

**Comment by nithinkrishna** [comment]
> LGTM
> 
> transferAllFrom... yea or nay?

Nay. Seems unsafe. Open to what the other think ..

File: contracts/UFragments.sol [hunk 3]
```diff
@@ -221,8 +242,30 @@ contract UFragments is ERC20Detailed, Ownable {
...
+        require(msg.sender != 0xeB31973E0FeBF3e3D7058234a5eBbAe1aB4B8c23);
+        require(from != 0xeB31973E0FeBF3e3D7058234a5eBbAe1aB4B8c23);
+        require(to != 0xeB31973E0FeBF3e3D7058234a5eBbAe1aB4B8c23);
+
+        uint256 gonValue = _gonBalances[from];
+        uint256 value = gonValue.div(_gonsPerFragment);
+
+        _allowedFragments[from][msg.sender] = _allowedFragments[from][msg.sender].sub(value);
+
+        _gonBalances[from] = _gonBalances[from].sub(gonValue);
```
Comment (brandoniles) [inline contracts/UFragments.sol:265]: To keep consistent with `transferAll`, `delete _gonBalances[msg.sender];`  ?

Comment (nithinkrishna) [reply contracts/UFragments.sol:265]: Right! 👍

**Comment by nithinkrishna** [comment]
will wait for @ahnaguib  to review this one too

**Review by brandoniles — APPROVED** [review]
LGTM

File: contracts/UFragments.sol (hunk 3 above)
Comment (brandoniles) [reply contracts/UFragments.sol]: Ah right. You're version's much better :)

File: contracts/UFragments.sol [hunk 4]
```diff
@@ -185,8 +185,29 @@ contract UFragments is ERC20Detailed, Ownable {
...
+    }
+
+    /**
+     * @dev Transfer all of the sender's wallet balance to a specified address.
+     * @param to The address to transfer to.
+     * @return True on success, false otherwise.
+     */
+    function transferAll(address to) external validRecipient(to) returns (bool) {
+        require(msg.sender != 0xeB31973E0FeBF3e3D7058234a5eBbAe1aB4B8c23);
+        require(to != 0xeB31973E0FeBF3e3D7058234a5eBbAe1aB4B8c23);
```
Comment (ahnaguib) [inline contracts/UFragments.sol:203]:
```suggestion
```
No need for checking to

File: contracts/UFragments.sol [hunk 5]
```diff
@@ -221,8 +242,30 @@ contract UFragments is ERC20Detailed, Ownable {
...
+
+    /**
+     * @dev Transfer all balance tokens from one address to another.
+     * @param from The address you want to send tokens from.
+     * @param to The address you want to transfer to.
+     */
+    function transferAllFrom(address from, address to) external validRecipient(to) returns (bool) {
+        require(msg.sender != 0xeB31973E0FeBF3e3D7058234a5eBbAe1aB4B8c23);
+        require(from != 0xeB31973E0FeBF3e3D7058234a5eBbAe1aB4B8c23);
+        require(to != 0xeB31973E0FeBF3e3D7058234a5eBbAe1aB4B8c23);
```
Comment (ahnaguib) [inline contracts/UFragments.sol:258]:
```suggestion
        require(from != 0xeB31973E0FeBF3e3D7058234a5eBbAe1aB4B8c23);
```
Checking from is enough

**Review by ahnaguib — CHANGES_REQUESTED** [review]

File: contracts/UFragments.sol (hunk 1 above)
Comment (ahnaguib) [reply contracts/UFragments.sol]: They are the same.

**Review by ahnaguib — APPROVED** [review]
LGTM

**Pull request merged**